contracts/lottery/src/lib.rs

```diff
@@ -0,0 +1,609 @@
+/// - `thresholds` - Defined thresholds with prizes
+fn count_total_prizes_percentage(
+    winners: &Map<u32, Vec<Address>>,
+    thresholds: &Map<u32, u32>,
+) -> u32 {
+    winners
+        .iter()
+        .fold(0u32, |acc, (threshold_number, _)| {
+            let threshold_percentage = thresholds.get(threshold_number).unwrap();
+            let winners_count = winners.get(threshold_number).unwrap().len();
+            acc.add(threshold_percentage * winners_count)
+        })
+}
+
+/// Recalculates new thresholds in case total prizes percentage is above 100%
+///
+/// # Arguments
+/// 
+/// - `winners` - A map containing winning players with a number of correctly selected numbers
+/// - `thresholds` - Defined thresholds with prizes, will be updated
+/// in case total_prizes_percentage > 100
+/// - `total_prizes_percentage` - Total prizes percentage
+fn recalculate_new_thresholds(
+    winners: &Map<u32, Vec<Address>>,
+    thresholds: &mut Map<u32, u32>,
+    total_prizes_percentage: u32,
+) {
+    if total_prizes_percentage > 100 {
+        for threshold_number in thresholds.keys() {
+            if winners.contains_key(threshold_number) {
+                let winners_count = winners.get(threshold_number).unwrap().len();
+                let threshold_precentage = thresholds.get(threshold_number).unwrap();
+                let val =
+                    winners_count * threshold_precentage * 100 / total_prizes_percentage;
+                thresholds.set(threshold_number, val / winners_count);
+            } else {
+                thresholds.remove(threshold_number);
+            }
+        }
+    }
+}
+
+/// Counts properly selected numbers for a given ticket
+///
+/// # Returns
+/// 
+/// Counted matches
+/// 
+/// # Arguments
+/// 
+/// - `drawn_numbers` - An array containing numbers selected in this lottery,
+/// - `player_ticket` - Numbers selected by a player
+fn count_matches(drawn_numbers: &Vec<u32>, player_ticket: &LotteryTicket) -> u32 {
+    drawn_numbers
+        .iter()
+        .filter(|x| player_ticket.contains(x))
+        .count() as u32
+}
+
+/// Randomly draw numbers within a given range (1, max_range).
+/// Ensures that there are no duplicates
+///
+/// # Returns
+/// 
+/// A list of randomly selected number
+/// 
+/// # Arguments
+///
+/// - `env` - The environment for this contract.
+/// - `max_range` - Right boundary of the range players will select numbers from (1, max_range)
+/// - `number_of_numbers` - Number of numbers possible to select by players
+/// - `random_seed` - A seed provided by the admin, that will be combined with other environment elements
+fn draw_numbers(env: &Env, max_range: u32, number_of_numbers: u32, random_seed: u64) -> Vec<u32> {
+    let mut numbers = Vec::new(&env);
+    for n in 0..number_of_numbers {
+        let new_seed = random_seed + n as u64;
+        let mut rand = SmallRng::seed_from_u64(new_seed);
```

Comment:
Also, in the latest version, looks like we can use env.prng() for accesing host pseudo randomness generators. Probably its going to be more efficient and we can remove a dependency. Just created a new issue for fixing other contracts: https://github.com/eigerco/nebula/issues/84

Reply:
Is there a way to initialize seed as we are doing it now when using `env.prng()`? I can only see either by integer literal or an 32 byte array. Not too convenient in our case...

Reply:
Not as pretty as its now ...

```rust
env.prng().seed(Bytes::from_slice(&env, 100u64.to_be_bytes().as_slice()));
```

Reply:
Have you tested this? I'm getting panic:
```
thread 'test::correct_lottery_results_are_returned' panicked at 'HostError: Error(Value, UnexpectedSize)

Event log (newest first):
   0: [Diagnostic Event] topics:[error, Error(Value, UnexpectedSize)], data:"escalating error to panic"
   1: [Diagnostic Event] topics:[error, Error(Value, UnexpectedSize)], data:["contract call failed", play_lottery, [666]]
   2: [Failed Diagnostic Event (not emitted)] contract:0499b1a330d4a90d62c7cc2e466b2b7fb43b1563db3591396cd4ad408fb8795c, topics:[error, Error(Value, UnexpectedSize)], data:"caught error from function"
   3: [Failed Diagnostic Event (not emitted)] contract:0499b1a330d4a90d62c7cc2e466b2b7fb43b1563db3591396cd4ad408fb8795c, topics:[error, Error(Value, UnexpectedSize)], data:"escalating error to panic"
   4: [Failed Diagnostic Event (not emitted)] contract:0499b1a330d4a90d62c7cc2e466b2b7fb43b1563db3591396cd4ad408fb8795c, topics:[error, Error(Value, UnexpectedSize)], data:["Unexpected size of BytesObject in prng_reseed", 8]
   ...
```

Reply:
There's another problem with `prng()` - all tests that are testing the helper functions of the contract and don't initialize it now fail because of this:
`1: [Diagnostic Event] topics:[error, Error(Context, MissingValue)], data:"no contract running"`

Reply:
I've managed to solve those problems, but I'm entirely happy from it, maybe you would have better idea. Basically I've added a `RandomGeneratorTrait` which is used in the `draw_numbers` function. There are now 2 different implementations of it - one in the contract code and one in tests:

lib.rs:
```rust
struct RandomNumberGenerator;

trait RandomGeneratorTrait {
    fn new(env: &Env, seed: u64) -> Self;
    fn number(&mut self, env: &Env, max_range: u32) -> u32;
}

impl RandomGeneratorTrait for RandomNumberGenerator {
    fn new(env: &Env, seed: u64) -> Self {
        let arr = [[0; 8], [0; 8], [0; 8], seed.to_be_bytes()].concat();
        env.prng().seed(Bytes::from_slice(&env, &arr.as_slice()));
        RandomNumberGenerator{}
    }

    fn number(&mut self, env: &Env, max_range: u32) -> u32 {
        env.prng().u64_in_range(1..=max_range as u64) as u32
    }
}

fn draw_numbers<T: RandomGeneratorTrait>(env: &Env, max_range: u32, number_of_numbers: u32, random_seed: u64) -> Vec<u32> {
    let mut numbers = Vec::new(&env);
    for n in 0..number_of_numbers {
        let new_seed = random_seed + n as u64;
        let mut random_generator = T::new(env, new_seed);
        loop {
            // draw a number so many times until a new unique number is found
            let drawn_number = random_generator.number(env, max_range.clone());
            if !numbers.contains(drawn_number) {
                numbers.push_back(drawn_number);
                break;
            }
        }
    }
    numbers
}
```

test.rs:
```rust
struct TestRandomNumberGenerator(SmallRng);

impl RandomGeneratorTrait for TestRandomNumberGenerator {
    fn new(_: &Env, seed: u64) -> Self {
        TestRandomNumberGenerator(SmallRng::seed_from_u64(seed))
    }

    fn number(&mut self, _: &Env, max_range: u32) -> u32 {
        self.0.gen_range(1..max_range)
    }
}
```
Now whenever we want to call this function directly in tests, we have to do something like:
```rust
let result = draw_numbers::<TestRandomNumberGenerator>(&env, 50, 5, 666);
```
WDYT?

Reply:
I like it ! Dont see any reasons why to not use it :) . This could be also needed in other tests, i would let it here but keep an eye in order to re-use it. Each day that passes i feel we need a dedicated crate for testing tools. Awesome work BTW !